File: 238_ProductOfArrayExceptSelf.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        result = [1] * len(nums)
        l, r = 1, 1
        for i in range(len(nums)):
            result[i] *= l
            l *= nums[i]
            result[len(nums) - i - 1] *= r
            r *= nums[len(nums) - i - 1]

        return result

File: 238.ProductofArrayExceptSelf.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int l = nums.length;
        int[] result = new int[l];
        int tmp = 1;
        for (int i = 0; i < l; i += 1) {
            result[i] = tmp;
            tmp *= nums[i];
        }
        tmp = 1;
        for (int i = l - 1; i >= 0; i -= 1) {
            result[i] *= tmp;
            tmp *= nums[i];
        }
        return result;
    }
}